bigchaindb/tep/upsert_validator.py

```diff
@@ -0,0 +1,181 @@
+import json
+
+from bigchaindb.tep.transactional_election import TransactionalElection
+from bigchaindb.tendermint.lib import BigchainDB
+from bigchaindb.models import Transaction
+from bigchaindb.tendermint.utils import (GO_AMINO_ED25519,
+                                         key_from_base64,)
+from bigchaindb.common.crypto import (key_pair_from_ed35519_key,
+                                      public_key_from_ed35519_key)
+
+
+class UpsertValidator(TransactionalElection):
+
+    version = '1.0'
+    name = 'upsert-validator'
+
+    def __init__(self, bigchain=None):
+        if not bigchain:
+            bigchain = BigchainDB()
+
+        self.bigchain = bigchain
+
+    def _new_election_object(self, args):
+        new_election_object = {
+            'type': 'election',
+            'name': self.name,
+            'version': self.version,
+            'args': args
+        }
+        return new_election_object
+
+    def _recipients(self):
+        """Convert validator dictionary to a recipient list for `Transaction`"""
+
+        recipients = []
+        for public_key, voting_power in self._validators().items():
+            recipients.append(([public_key], voting_power))
+
+        return recipients
+
+    def _validators(self):
+        """Return a dictionary of validators with key as `public_key` and
+           value as the `voting_power`
+        """
+
+        validators = {}
+        for validator in self.bigchain.get_validators():
+            if validator['pub_key']['type'] == GO_AMINO_ED25519:
+                public_key = public_key_from_ed35519_key(key_from_base64(validator['pub_key']['value']))
+                validators[public_key] = validator['voting_power']
+            else:
+                # TODO: use appropriate exception
+                raise NotImplementedError
+
+        return validators
+
+    def _get_vote(self, tx_election, public_key):
+        """For a given `public_key` return the `input` and `amount` as
+           voting power
+        """
+
+        for i, tx_input in enumerate(tx_election.to_inputs()):
+            if tx_input.owners_before == [public_key] and \
+               tx_election.output[i].public_keys == [public_key]:
+                return ([tx_input], tx_election.output[i].amount)
+
+        return ([], 0)
+
+    def _is_same_topology(current_topology, election_topology):
+
+        voters = {}
+        for voter in election_topology:
+            [public_key] = voter.public_keys
```

Comment:
There is a special case of `len(voter.public_keys) > 1`. But then if we handle it here, a better name for the function might be `_is_valid_election_topology`.

Reply:
Didn't understanding the comment, could you please re-phrase?

Reply:
I mean, in case `len(voter.public_keys) > 1`, it makes sense return `False` here instead of raising an unpacking exception.